File: contents/stacks_and_queues/code/cpp/stack.cpp
#include<iostream>
#include<cassert>
#include<memory>

namespace my {
    /**
     * implementation using linked list
     * [value][next] -> [value][next] -> ... -> [value][next]
     * (top Node)      (intermediat Nodes)
     * left most Node represents top element of stack
     */
    template<typename T>
    struct Node {
        /**
        * next: will store right Node address
        */
        T value;
        std::unique_ptr<Node<T>> next;
        Node(const T& V) : value(V) { }
    };

    template<typename T>
    class stack {
    private:
        /**
         * top_pointer: points to left most node
         * count: keeps track of current number of elements present in stack
         */
        std::unique_ptr<Node<T>> top_pointer;
        size_t count;
    public:
        stack() : count(0ULL) { }

        void push(const T& element) {
            auto new_node = std::make_unique<Node<T>>(element);
            new_node->next = std::move(top_pointer);
            top_pointer = std::move(new_node);
            count = count + 1;
        }

        void pop() {
            if (count > 0) {
                top_pointer = std::move(top_pointer->next);
                count = count - 1;
            }
        }

        T& top() {
            assert(count > 0 and "calling top() on an empty stack");
            return top_pointer->value;
        }
        // returning mutable reference can very be usefull if someone wants to modify top element

        T const& top() const {
            assert(count > 0 and "calling top() on an empty stack");
            return top_pointer->value;
        }

        size_t size() const { return count; }

        bool empty() const { return count == 0; }

        ~stack() {
            while (top_pointer.get() != nullptr) {
                top_pointer = std::move(top_pointer->next);
            }
        }
    };
}

int main() {
  my::stack<int> intStack;

  intStack.push(4);
  intStack.push(5);
  intStack.push(9);

  int topElement = intStack.top();
  intStack.pop();
  std::cout << topElement << '\n';
  std::cout << intStack.size() << '\n';
  std::cout << intStack.top() << '\n';
  return 0;
}

File: contents/stacks_and_queues/code/csharp/Stack.cs
using System;
using System.Collections.Generic;

namespace my
{
    /**
     * implementation using linked list
     * [value][next] -> [value][next] -> ... -> [value][next]
     * (top Node)      (intermediat Nodes)
     * left most Node represents top element of stack
     */
    public class Node {
        public Node next;
        public object data;
    }

    public class Stack {
        private Node head;
        public int size;

        public Stack() {
            head = new Node();
            size = 0;
        }
        public void Push(object value) {
            Node newNode = new Node();
            newNode.data = value;
            newNode.next = head.next;
            head.next = newNode;
            size++;
        }

        public void Pop() {
            if (size != 0) {
                head.next = head.next.next;
                size--;
            }
            else {
                Console.WriteLine("No element to remove.");
            }
        }

        public object Top() {
            return head.next.data;
        }

        public int Size() {
            return size;
        }

        public bool Empty() {
            return size == 0;
        }

    }

    class Test {
        static void Main(string[] args) {

            Console.WriteLine("Stack:");
            Stack intStack = new Stack();

            intStack.Push(4);
            intStack.Push(5);
            intStack.Push(9);

            Console.Write("Size: ");
            Console.WriteLine(intStack.Size());
            Console.Write("Top: ");
            Console.WriteLine(intStack.Top());

            intStack.Pop();

            Console.Write("Size: ");
            Console.WriteLine(intStack.Size());
            Console.Write("Top: ");
            Console.WriteLine(intStack.Top());
        }
    }
}
